runtime/near-vm-logic/src/logic.rs

```diff
@@ -109,88 +133,162 @@ impl<'a> VMLogic<'a> {
+    memory_get!(u32, memory_get_u32);
+    memory_get!(u16, memory_get_u16);
+    memory_get!(u8, memory_get_u8);
 
     /// Reads an array of `u64` elements.
-    fn memory_get_array_u64(
-        memory: &dyn MemoryLike,
-        offset: u64,
-        num_elements: u64,
-    ) -> Result<Vec<u64>> {
+    fn memory_get_vec_u64(&mut self, offset: u64, num_elements: u64) -> Result<Vec<u64>> {
         let memory_len = num_elements
             .checked_mul(size_of::<u64>() as u64)
             .ok_or(HostError::MemoryAccessViolation)?;
-        let data = Self::memory_get(memory, offset, memory_len)?;
-        Ok(data
-            .chunks(size_of::<u64>())
-            .map(|buf| {
-                assert_eq!(buf.len(), size_of::<u64>());
-                let mut array = [0u8; size_of::<u64>()];
-                array.copy_from_slice(buf);
-                u64::from_le_bytes(array)
-            })
-            .collect())
+        let data = self.memory_get_vec(offset, memory_len)?;
+        let mut res = vec![0u64; num_elements as usize];
+        byteorder::LittleEndian::read_u64_into(&data, &mut res);
+        Ok(res)
     }
 
-    fn read_memory_u32(memory: &dyn MemoryLike, ptr: u64) -> Result<u32> {
-        let mut slice = [0u8; size_of::<u32>()];
-        let buf = Self::memory_get(memory, ptr, size_of::<u32>() as u64)?;
-        slice.copy_from_slice(&buf);
-        Ok(u32::from_le_bytes(slice))
+    fn get_vec_from_memory_or_register(&mut self, offset: u64, len: u64) -> Result<Vec<u8>> {
+        if len != std::u64::MAX {
+            self.memory_get_vec(offset, len)
+        } else {
+            self.internal_read_register(offset)
+        }
     }
 
-    fn get_from_memory_or_register(
-        memory: &dyn MemoryLike,
-        registers: &HashMap<u64, Vec<u8>>,
-        offset: u64,
-        len: u64,
-    ) -> Result<Vec<u8>> {
-        if len != std::u64::MAX {
-            Self::memory_get(memory, offset, len)
+    fn memory_set_slice(&mut self, offset: u64, buf: &[u8]) -> Result<()> {
+        self.gas_counter.pay_base(write_memory_base)?;
+        self.gas_counter.pay_per_byte(write_memory_byte, buf.len() as _)?;
+        self.try_fit_mem(offset, buf.len() as _)?;
+        self.memory.write_memory(offset, buf);
+        Ok(())
+    }
+
+    memory_set!(u128, memory_set_u128);
+
+    // #################
+    // # Registers API #
+    // #################
+
+    fn internal_read_register(&mut self, register_id: u64) -> Result<Vec<u8>> {
+        if let Some(data) = self.registers.get(&register_id) {
+            self.gas_counter.pay_base(read_register_base)?;
+            self.gas_counter.pay_per_byte(read_register_byte, data.len() as _)?;
+            Ok(data.clone())
+        } else {
+            Err(HostError::InvalidRegisterId.into())
+        }
+    }
+
+    fn internal_write_register(&mut self, register_id: u64, data: Vec<u8>) -> Result<()> {
+        self.gas_counter.pay_base(write_register_base)?;
```

Comment:
Minor thing, but we could also save cycles for calling counter if we will generate a method for each of the fields of ExtCosts instead (and even #inline it) instead of have to match it against variant each time.

Reply:
Yes, the runtime can be optimized after we finish non-optimization related features.